network/src/lib.rs

```diff
@@ -49,3 +49,35 @@ pub use tokio;
 
 /// Protocol version used by network protocol open
 pub type ProtocolVersion = String;
+
+/// Observe listen port occupancy
+pub fn observe_listen_port_occupancy(
+    _addrs: &[multiaddr::MultiAddr],
+) -> Result<(), std::io::Error> {
+    #[cfg(target_os = "linux")]
+    {
+        use multiaddr::Protocol;
+        use std::net::{IpAddr, Ipv4Addr, SocketAddr, TcpListener};
+
+        for addr in _addrs {
+            let mut iter = addr.iter().peekable();
+
+            let mut ip_addr: Option<SocketAddr> = None;
+
+            while iter.peek().is_some() {
+                let proto = iter.next().unwrap();
+
+                if let Protocol::Tcp(port) = proto {
+                    ip_addr = Some(SocketAddr::new(IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)), port));
```

Comment:
I think we should match `Protocol::Ip4` and use the matched ip address / port to build the `SocketAddr`

Reply:
No need to do this because after any port of this machine is bound, this port cannot be bound on any IP combination, whether it is the loopback address, the network card address, the all-zero address

Reply:
user may use multiple NICs in vm and multiple ckb process listen on the different ip address with same port, using `0.0.0.0` to check port availability will breaks this use case.

Reply:
You are right, it has been modified